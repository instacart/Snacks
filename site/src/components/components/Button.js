import React, {Fragment} from 'react'
import Playground from '../Playground'
import {Title2, Title3, SectionTitle, Subtitle, Body} from '../Typography'
import PropsDocs from '../PropsDocs'
import ButtonDocs from '!!react-docgen-loader!../../../../src/components/Buttons/Button'

export default () => (
  <Fragment>
    <Title2>Button</Title2>
    <Subtitle>
      This is our Button.
      There are many like it, but this one is ours.

      Buttons guide our users to take a certain action, like 'Add to Cart' or
      'Checkout'. The call to action should start with an active verb, and the
      style choice of the button should reflect the hierarchy and importance of
      the action.
    </Subtitle>
    <PropsDocs docs={ButtonDocs} />

    <Title3>Styles</Title3>
    <SectionTitle>Primary</SectionTitle>
    <Body>
      Primary buttons are for, you guessed it, the primary action. Don't use
      these lightly. Overuse on a single screen could hinder your main objective
      rather than help it.
    </Body>

    <Playground>
      {`
        <Button snacksStyle="primary">
          Primary Button
        </Button>
      `}
    </Playground>

    <SectionTitle>Secondary</SectionTitle>
    <Body>
      Secondary buttons are for your second most important action.
      A good use is as a cancel button when asking the user a question.
    </Body>

    <Playground>
      {`
        <Button snacksStyle="secondary">
          Secondary Button
        </Button>
      `}
    </Playground>

    <SectionTitle>Flat</SectionTitle>
    <Body>
      The flat style can be used when you want a button that looks like a link.
      A typical solution is to use an anchor tag to do this, however this approach
      has issues such as not having consistent padding to a button and there are
      accessibility concerns (links should navigate, buttons should cause actions).
    </Body>

    <Playground>
      {`
        <Button snacksStyle="flat">
          Flat Button
        </Button>
      `}
    </Playground>

    <SectionTitle>Coupon</SectionTitle>
    <Body>
      Note that coupon buttons aren't affected by a Snacks theme.
    </Body>

    <Playground>
      {`
        <Button snacksStyle="coupon">
          Coupon Button
        </Button>
        `}
    </Playground>

    <SectionTitle>Disabled</SectionTitle>
    <Body>
      Disabled buttons are to show the user there is a future action they can
      take, but they must first accomplish another task.
    </Body>

    <Playground>
      {`
        <Button disabled snacksStyle="primary">
          Disabled Button
        </Button>
        `}
    </Playground>

    <SectionTitle>Inverted colors</SectionTitle>
    <Body>
      Button colors can be inverted for darker backgrounds.
    </Body>

    <Playground>
      {`
        <div style={{padding: spacing.MD, backgroundColor: colors.GREEN_500}}>
          <Button inverted snacksStyle="primary" style={{marginRight: spacing.SM}}>
            Primary Button
          </Button>
          <Button inverted snacksStyle="secondary" style={{marginRight: spacing.SM}}>
            Secondary Button
          </Button>
        </div>
      `}
    </Playground>

    <Title3>Sizes</Title3>
    <Body>
      Buttons can also be set to a variety of sizes.
    </Body>
    <Playground>
      {`
        <div>
          <Button size="large" style={{marginBottom: spacing.SM, display: 'block'}}>
            Large Button
          </Button>
          <Button size="standard" style={{marginBottom: spacing.SM, display: 'block'}}>
            Standard Button
          </Button>
          <Button size="small" style={{marginBottom: spacing.SM, display: 'block'}}>
            Small Button
          </Button>
          <Button size="tiny" style={{marginBottom: spacing.SM, display: 'block'}}>
            Tiny Button
          </Button>
        </div>
      `}
    </Playground>

    <Title3>With icons</Title3>
    <Body>
      Buttons can be rendered with an icon to the left or right of the primary
      content.
    </Body>

    <Playground>
      {`
        <div>
          <Button iconPosition="right" icon={<Icon name="arrowRightSmallBold" />} snacksStyle="secondary" style={{marginBottom: spacing.SM, display: 'block'}}>
            View 24 more
          </Button>
          <Button icon="cart" style={{marginBottom: spacing.SM, display: 'block'}}>
            Cart
          </Button>
          <Button icon={<Icon name="deals" />} snacksStyle="coupon" size="small" style={{marginBottom: spacing.SM, display: 'block'}}>
            Save $5.00
          </Button>
        </div>
      `}
    </Playground>
  </Fragment>
)
