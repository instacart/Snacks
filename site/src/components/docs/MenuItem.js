import React, {Fragment} from 'react'
import Playground from '../Playground'

export default () => (

  <Fragment>
    MenuItem should always be wrapped in parent Menu or DropdownMenu. And onClick should be accessed through the parent's onSelect callback.
  </Fragment>

)
