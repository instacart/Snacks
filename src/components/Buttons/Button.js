import React from 'react'
import PropTypes from 'prop-types'
import Radium from 'radium'
import Icon from '../Icon/Icon'
import withTheme from '../../styles/themer/withTheme'
import { themePropTypes } from '../../styles/themer/utils'
import { spacing, colors } from '../../styles'
import { darken } from '../../utils'

const noop = () => {} // eslint-disable-line no-empty-function

const baseStyles = {
  touchAction: 'manipulation',
  cursor: 'pointer',
  border: '1px solid transparent',
  borderTopRightRadius: 4,
  borderBottomRightRadius: 4,
  borderBottomLeftRadius: 4,
  borderTopLeftRadius: 4,
  fontWeight: 600,
  whiteSpace: 'nowrap',
  userSelect: 'none',
  WebkitFontSmoothing: 'antialiased',

  // Reset unusual Firefox-on-Android default style;
  // see https://github.com/necolas/normalize.css/issues/214
  backgroundImage: 'none',

  // Takes care of the heavier default font-weights in firefox
  MozOsxFontSmoothing: 'grayscale',

  // Ensures any icons stay vertically centered
  display: 'inline-flex',
  alignItems: 'center'
}

const sizeStyles = {
  tiny: {
    ...spacing.PADDING_X_XS,
    height: '24px',
    fontSize: '11px',
    textTransform: 'uppercase'
  },

  small: {
    ...spacing.PADDING_X_XS,
    fontSize: '14px',
    height: '32px'
  },

  standard: {
    ...spacing.PADDING_X_SM,
    fontSize: '16px',
    height: '40px'
  },

  large: {
    ...spacing.PADDING_X_MD,
    fontSize: '18px',
    height: '48px'
  }
}

const linkStyles = {
  display: 'flex',
  alignItems: 'center',
  boxSizing: 'border-box'
}

const getSnacksStyles = props => {
  const { action, actionHover, primaryBackground } = props.snacksTheme.colors
  const actionActive = darken(actionHover, 3)

  const disabled = {
    backgroundColor: colors.GRAY_74,
    color: colors.WHITE,
    cursor: 'not-allowed'
  }

  const activeAndFocus = {
    primary: {
      base: {
        backgroundColor: actionActive
      },
      inverted: {
        backgroundColor: primaryBackground,
        opacity: '0.8'
      }
    },
    secondary: {
      base: {
        color: actionActive,
        border: `1px solid ${actionActive}`
      },
      inverted: {
        color: primaryBackground,
        border: `1px solid ${primaryBackground}`,
        opacity: '0.8'
      }
    },
    flat: {
      color: actionActive,
    },
    coupon: {
      border: `1px dashed ${colors.RED_700}`,
      color: colors.RED_700
    }
  }

  return {
    primary: {
      base: {
        backgroundColor: action,
        color: primaryBackground,

        ':hover': {
          backgroundColor: actionHover
        },
        ':active': activeAndFocus.primary.base,
        ':focus': activeAndFocus.primary.base
      },

      inverted: {
        backgroundColor: primaryBackground,
        border: `1px solid ${action}`,
        color: action,

        ':hover': {
          backgroundColor: primaryBackground,
          opacity: '0.9'
        },
        ':active': activeAndFocus.primary.inverted,
        ':focus': activeAndFocus.primary.inverted
      },

      disabled
    },

    secondary: {
      base: {
        backgroundColor: 'transparent',
        color: action,
        border: `1px solid ${action}`,

        ':hover': {
          color: actionHover,
          border: `1px solid ${actionHover}`
        },
        ':active': activeAndFocus.secondary.base,
        ':focus': activeAndFocus.secondary.base
      },

      inverted: {
        color: primaryBackground,
        border: `1px solid ${primaryBackground}`,

        ':hover': {
          color: primaryBackground,
          border: `1px solid ${primaryBackground}`,
          opacity: '0.9'
        },
        ':active': activeAndFocus.secondary.inverted,
        ':focus': activeAndFocus.secondary.inverted
      },

      disabled
    },

    flat: {
      base: {
        backgroundColor: 'transparent',
        color: action,

        ':hover': {
          color: actionHover,
        },
        ':active': activeAndFocus.flat,
        ':focus': activeAndFocus.flat
      },

      disabled: {
        backgroundColor: 'transparent',
        color: colors.GRAY_74,
        cursor: 'not-allowed'
      }
    },

    coupon: {
      base: {
        backgroundColor: colors.WHITE,
        color: colors.RED_500,
        border: `1px dashed ${colors.RED_500}`,

        ':hover': {
          border: `1px dashed ${colors.RED_600}`,
          color: colors.RED_600
        },
        ':active': activeAndFocus.coupon,
        ':focus': activeAndFocus.coupon
      },

      disabled
    }
  }
}

const Button = props => {
  const snacksStyles = getSnacksStyles(props)

  const ElementType = props.href ? 'a' : 'button'

  const finalProps = {
    disabled: props.disabled,
    tabIndex: props.tabIndex,
    type: props.type,
    style: [
      baseStyles,
      sizeStyles[props.size],
      snacksStyles[props.snacksStyle][props.disabled ? 'disabled' : 'base'],
      props.inverted && snacksStyles[props.snacksStyle].inverted,
      ElementType === 'a' && linkStyles,
      props.style
    ],
    onClick: e => {
      if (props.disabled) {
        e.preventDefault()
        return
      }

      props.onClick(e, props)
    },
    onMouseDown: e => {
      if (props.disabled) {
        e.preventDefault()
        return
      }

      props.onMouseDown(e, props)
    },
    ...props.elementAttributes
  }
  if (props.href) {
    finalProps.href = props.href
  }

  const icon =
    typeof props.icon === 'string' ? <Icon name={props.icon} /> : props.icon

  if (icon && props.iconPosition === 'left') {
    return (
      <ElementType {...finalProps}>
        {icon}
        <span style={{ ...spacing.MARGIN_RIGHT_XS }} />
        {props.children}
      </ElementType>
    )
  }

  if (icon && props.iconPosition === 'right') {
    return (
      <ElementType {...finalProps}>
        {props.children}
        <span style={{ ...spacing.MARGIN_LEFT_XS }} />
        {icon}
      </ElementType>
    )
  }

  return <ElementType {...finalProps}>{props.children}</ElementType>
}

Button.propTypes = {
  /** Size of the button. */
  size: PropTypes.oneOf(['tiny', 'small', 'standard', 'large']),

  /** Snacks button variants. */
  snacksStyle: PropTypes.oneOf(['primary', 'secondary', 'flat', 'coupon']),

  /** Optional style overrides. */
  style: PropTypes.object,

  /** Whether or not the button is disabled. */
  disabled: PropTypes.bool,

  /** Sets the HTML type attribute on the element. */
  type: PropTypes.oneOf(['button', 'reset', 'submit']),

  /**
    An optional icon. Can be a an icon name
    or a Snacks `Icon` component.
  */
  icon: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),

  /** Where the icon is positioned relative to primary text content. */
  iconPosition: PropTypes.oneOf(['left', 'right']),

  /** Callback that fires when the button is clicked. */
  onClick: PropTypes.func,

  /** Callback that fires on mouse down. */
  onMouseDown: PropTypes.func,

  /** Reverses colors. Use for rendering buttons on dark backgrounds. */
  inverted: PropTypes.bool,

  /** Sets the HTML `tabindex` attribute. */
  tabIndex: PropTypes.number,

  /**
    If href is provided, `Button` will render as an anchor
    tag with the supplied href value.
  */
  href: PropTypes.string,

  /** Text content for the button. */
  children: PropTypes.node,

  /** Any additonal props to add to the element (e.g. data attributes). */
  elementAttributes: PropTypes.object,

  /** Snacks theme attributes provided by `Themer` */
  snacksTheme: themePropTypes
}

Button.defaultProps = {
  size: 'standard',
  snacksStyle: 'primary',
  disabled: false,
  type: 'button',
  iconPosition: 'left',
  onClick: noop,
  onMouseDown: noop,
  inverted: false,
  elementAttributes: {}
}

export default withTheme(Radium(Button))
