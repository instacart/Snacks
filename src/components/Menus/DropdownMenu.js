import React from 'react'
import { findDOMNode } from 'react-dom'
import PropTypes from 'prop-types'
import Menu from './Menu'
import Slide from '../Transitions/Slide'
import Fade from '../Transitions/Fade'

const styles = {
  menuContainer: {
    position: 'absolute',
    width: '100%',
    display: 'block',
    pointerEvents: 'none',
  },
  menuContainerClosed: {
    zIndex: -9000,
    pointerEvents: 'none',
  },
  menuContainerOpen: {
    zIndex: 9000,
    pointerEvents: 'inherit',
  },
  transitionContainer: {
    paddingTop: '2px',
    paddingRight: '5px',
    paddingBottom: '5px',
    paddingLeft: '5px',
    width: '100%',
    transform: 'translateX(-5px)',
  },
}

class DropdownMenu extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    /** MenuItems or Divider */
    children: PropTypes.node.isRequired,
    /** React component or HTML that will be used as the dropdown trigger. ie. Button or Icon */
    triggerElement: PropTypes.element,
    /** Open and close menu manually */
    open: PropTypes.bool,
    /** Callback fired on dropdown close */
    onClose: PropTypes.func,
    /** Callback fired on click */
    onClick: PropTypes.func,
    /** Callback fired on dropdown open */
    onOpen: PropTypes.func,
    /** Callback function fired on key down */
    onKeyDown: PropTypes.func,
    /** Requesting to open/close for controlled open component. */
    onRequestChange: PropTypes.func,
    /** Callback function fired when a MenuItem is selected */
    onSelect: PropTypes.func,
    /**
     * Optional style overrides merged into emotion css
     *
     * @deprecated
     * This prop exists for backwards compatibility and will be
     * removed in a future version
     */
    style: PropTypes.shape({}),
    /**
     * Customize style of menu parent
     *
     * @deprecated
     * This prop exists for backwards compatibility and will be
     * removed in a future version
     */
    menuContainerStyle: PropTypes.shape({}),
    /** Customize style of menu parent */
    menuContainerClassName: PropTypes.string,
    /** Props passed to Menu component */
    menuProps: PropTypes.shape({}),
  }

  static defaultProps = {
    open: null,
  }

  state = {
    open: false,
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.open !== null && this.props.open !== nextProps.open) {
      if (nextProps.open) {
        this.open()
      } else {
        this.close()
      }
    }
  }

  handleSelect = (e, option) => {
    const { onSelect, onRequestChange } = this.props

    if (this.controlledOpen()) {
      onRequestChange && onRequestChange(false)
    } else {
      // if multi then do not close
      this.close()
    }

    findDOMNode(this.trigger).focus()
    onSelect && onSelect(e, option)
  }

  handleClick = event => {
    event.preventDefault()
    const { open, onClick, onRequestChange } = this.props

    if (this.controlledOpen()) {
      onRequestChange && onRequestChange(!open)
      return
    }

    if (!this.state.open) {
      this.open()
    } else {
      this.close()
    }
    onClick && onClick(event)
  }

  handleMenuBlur = () => {
    const { onRequestChange } = this.props
    if (this.controlledOpen()) {
      onRequestChange && onRequestChange(false)
    } else {
      this.close()
    }
  }

  handleKeyDown = event => {
    const { onRequestChange } = this.props
    switch (event.key) {
      case 'Enter':
        if (!event.defaultPrevented && this.controlledOpen()) {
          event.preventDefault()
          onRequestChange && onRequestChange(true)
        } else {
          event.preventDefault()
          if (this.state.open) {
            this.close()
          } else {
            this.open()
          }
        }
        break
      case 'Escape':
        event.preventDefault()

        if (this.controlledOpen()) {
          onRequestChange && onRequestChange(false)
        } else {
          this.state.open && this.close()
        }
        break
    }
  }

  controlledOpen() {
    return this.props.open !== null
  }

  open() {
    const { onOpen } = this.props
    this.setState({ open: true }, () => {
      this.menu.focus()
      onOpen && onOpen()
    })
  }

  close() {
    const { onClose } = this.props
    this.setState({ open: false }, () => {
      this.menu.blur()
      onClose && onClose()
    })
  }

  renderTriggerElement() {
    const { triggerElement } = this.props
    const { open } = this.state
    if (triggerElement) {
      return React.cloneElement(triggerElement, {
        ref: node => {
          this.trigger = node
          const { ref } = triggerElement
          if (typeof ref === 'function') {
            ref(node)
          }
        },
        onMouseDown: this.handleClick,
        elementAttributes: {
          'aria-haspopup': true,
          'aria-expanded': open,
        },
        'aria-haspopup': true,
        'aria-expanded': open,
      })
    }
  }

  render() {
    const {
      className,
      style,
      children,
      menuProps,
      menuContainerStyle,
      menuContainerClassName,
    } = this.props

    const isOpen = this.state.open

    return (
      <div onKeyDown={this.handleKeyDown} css={style} className={className}>
        <div css={{ position: 'relative' }}>{this.renderTriggerElement()}</div>
        <div
          className={menuContainerClassName}
          css={[
            styles.menuContainer,
            menuContainerStyle,
            isOpen && styles.menuContainerOpen,
            !isOpen && styles.menuContainerClosed,
          ]}
        >
          <Slide in={isOpen} axis="y" css={styles.transitionContainer} offset={30}>
            <Fade in={isOpen} transitionTime={200}>
              <Menu
                ref={node => (this.menu = node)}
                ariaHidden={!isOpen}
                onBlur={this.handleMenuBlur}
                onSelect={this.handleSelect}
                {...menuProps}
              >
                {children}
              </Menu>
            </Fade>
          </Slide>
        </div>
      </div>
    )
  }
}

export default DropdownMenu
