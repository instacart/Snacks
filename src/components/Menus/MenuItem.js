import React from 'react'
import PropTypes from 'prop-types'
import Radium from 'radium'
import { colors } from '../../styles'
import Icon from '../Icon/Icon'
import spacing from '../../styles/spacing'

/* eslint jsx-a11y/no-noninteractive-tabindex: 0 */

const styles = {
  root: {
    backgroundColor: '#ffffff',
    ...spacing.PADDING_Y_SM,
    paddingRight: 0,
    paddingLeft: 0,
    ':hover': {
      cursor: 'pointer',
      backgroundColor: colors.GRAY_97,
    },
    ':focus': {
      backgroundColor: colors.GRAY_97,
    },
  },
  focus: {
    backgroundColor: colors.GRAY_97,
  },
  disabled: {
    color: colors.GRAY_74,
    backgroundColor: '#FFF',
    pointerEvents: 'none',
  },
  menuitem: {
    display: 'table',
  },
  iconContainer: {
    display: 'table-cell',
    verticalAlign: 'middle',
    paddingLeft: '16px',
  },
  label: {
    ...spacing.PADDING_X_SM,
    paddingTop: 0,
    paddingBottom: 0,
    whiteSpace: 'nowrap',
  },
  leftIconStyles: {
    fontSize: '22px',
  },
}

@Radium
class MenuItem extends React.Component {
  static propTypes = {
    /** Completely override the MenuItem rendering and create a custom MenuItem */
    children: PropTypes.node,
    /** Disable the MenuItem */
    disabled: PropTypes.bool,
    /** Focus the MenuItem */
    focus: PropTypes.bool,
    /** Index of MenuItem within Menu. Used for currentIndex */
    index: PropTypes.number,
    /** Text for the menu item */
    label: PropTypes.string.isRequired,
    /** Override styles of label */
    labelStyles: PropTypes.shape({}),
    /** Icon name or Icon component displayed left of label */
    leftIcon: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    /** Override styles for leftIcon */
    leftIconStyles: PropTypes.shape({}),
    /** Callback function fired when the menu item is click. Overriden by parent Menu or DropdownMenu */
    _onClick: PropTypes.func,
    /** Callback function fired when the menu item is focused. */
    onFocus: PropTypes.func,
    /** Used by menu to keep track of current focus index. */
    onMenuItemFocus: PropTypes.func,
    /** Whether or not to prevent default when menu item is clicked */
    preventDefault: PropTypes.bool,
    /** Role HTML attribute */
    role: PropTypes.string,
    /** Customize style of MenuItem */
    style: PropTypes.shape({}),
    /** Override tabIndex property */
    tabIndex: PropTypes.number,
    /** Whether or not to use use the tabIndex HTML attribute */
    useTabIndex: PropTypes.bool,
    /** Underlying value. Also, passed into _onClick function */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]).isRequired,
  }

  static defaultProps = {
    disabled: false,
    role: 'menuitem',
    useTabIndex: true,
    focus: false,
    preventDefault: true,
  }

  static isSnacksMenuItem = true

  componentDidUpdate(prevProps) {
    if (!prevProps.focus && this.props.focus) {
      setTimeout(() => {
        // Checking the ref exists in case this component
        // unmounts before the callback runs.
        this.menuItem && this.menuItem.focus()
      }, 0)
    }
  }

  handleClick = e => {
    const { disabled, index, _onClick, value, label, preventDefault } = this.props
    preventDefault && e.preventDefault()

    const option = { value, label }
    !disabled && _onClick && _onClick(e, option, index)
  }

  handleKeyDown = event => {
    switch (event.key) {
      case 'Enter':
        this.props.preventDefault && event.preventDefault()
        this.handleClick(event)
    }
  }

  handleFocus = event => {
    const { onFocus, onMenuItemFocus, index } = this.props
    onMenuItemFocus && onMenuItemFocus(index)
    onFocus && onFocus(index, event)
  }

  renderLeftIcon() {
    const { leftIcon, leftIconStyles } = this.props
    if (!leftIcon) {
      return
    }

    const iconComponent =
      typeof leftIcon === 'string' ? (
        <Icon name={leftIcon} style={[styles.leftIconStyles, leftIconStyles]} />
      ) : (
        leftIcon
      )

    return (
      <div style={styles.iconContainer}>
        <div style={{ display: 'flex' }}>{iconComponent}</div>
      </div>
    )
  }

  renderMenuItem() {
    const { children, label, labelStyles } = this.props

    if (children) {
      return children
    }

    return (
      <div style={styles.menuitem}>
        {this.renderLeftIcon()}
        <div style={[styles.label, labelStyles]}>{label}</div>
      </div>
    )
  }

  render() {
    const { disabled, role, style, tabIndex, useTabIndex } = this.props

    return (
      <div
        ref={node => (this.menuItem = node)}
        role={role}
        style={[styles.root, style, disabled && styles.disabled]}
        onClick={this.handleClick}
        onFocus={this.handleFocus}
        tabIndex={useTabIndex && !disabled ? tabIndex || 0 : undefined}
        onKeyDown={this.handleKeyDown}
      >
        {this.renderMenuItem()}
      </div>
    )
  }
}

export default MenuItem
