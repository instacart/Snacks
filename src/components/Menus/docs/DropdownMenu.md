DropdownMenu Example:

```js
<div style={{display: 'flex'}}>
  <div style={{marginRight: '20px'}}>
    <DropdownMenu triggerElement={<Button> Share </Button>}>
      <MenuItem label="Share via Email" value="email" leftIcon="emailFilled" />
      <MenuItem label="Share via Facebook" value="facebook" leftIcon="facebookFilled" leftIconStyles={{color: '#3c5a99'}} />
      <MenuItem label="Share via Twitter" value="twitter" leftIcon="twitterFilled" leftIconStyles={{color: '#1da1f2'}}/>
    </DropdownMenu>
  </div>
  <div>
    <DropdownMenu triggerElement={
      <Button> <Icon name="shareAndroid2Filled"/> </Button>
    }>
      <MenuItem label="Share via Email" value="email" leftIcon="emailFilled" />
      <MenuItem label="Share via Facebook" value="facebook" leftIcon="facebookFilled" leftIconStyles={{color: '#3c5a99'}} />
      <MenuItem label="Share via Twitter" value="twitter" leftIcon="twitterFilled" leftIconStyles={{color: '#1da1f2'}}/>
    </DropdownMenu>
  </div>
</div>
```
