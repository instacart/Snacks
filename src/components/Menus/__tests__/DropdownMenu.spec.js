import React         from 'react'
import renderer      from 'react-test-renderer'
import { StyleRoot } from 'radium'
import { mount }     from 'enzyme'
import { spy }       from 'sinon'
import DropdownMenu  from '../DropdownMenu'
import MenuItem      from '../MenuItem'
import Button        from '../../Buttons/Button'

it('renders DropdownMenu with icons and trigger correctly', () => {
  const tree = renderer.create(
    <StyleRoot>
      <div>
        <DropdownMenu triggerElement={<Button> Share </Button>}>
          <MenuItem label="Share via Email" value="email" leftIcon="emailFilled" />
          <MenuItem label="Share via Facebook" value="facebook" leftIcon="facebookFilled" leftIconStyles={{color: '#3c5a99'}} />
          <MenuItem label="Share via Twitter" value="twitter" leftIcon="twitterFilled" leftIconStyles={{color: '#1da1f2'}}/>
        </DropdownMenu>
      </div>
    </StyleRoot>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should call callbacks correctly', () => {
  const onClose = spy()
  const onOpen = spy()
  const onSelect = spy()

  const wrapper = mount(
    <DropdownMenu
      onClose={onClose}
      onOpen={onOpen}
      onSelect={onSelect}
      triggerElement={<Button> Share </Button>}
    >
      <MenuItem label="Share via Email" value="email" leftIcon="emailFilled" />
      <MenuItem label="Share via Facebook" value="facebook" leftIcon="facebookFilled" leftIconStyles={{color: '#3c5a99'}} />
      <MenuItem label="Share via Twitter" value="twitter" leftIcon="twitterFilled" leftIconStyles={{color: '#1da1f2'}}/>
    </DropdownMenu>
  )

  wrapper.find(Button).first().simulate('mouseDown')
  expect(onOpen.calledOnce).toBe(true)

  wrapper.find(MenuItem).first().simulate('click')
  expect(onSelect.calledOnce).toBe(true)
  expect(onClose.calledOnce).toBe(true)
})
