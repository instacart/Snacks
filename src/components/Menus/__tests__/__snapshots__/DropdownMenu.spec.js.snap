// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DropdownMenu with icons and trigger correctly 1`] = `
<div
  data-radium={true}
>
  <div>
    <div
      data-radium={true}
      onKeyDown={[Function]}
      style={undefined}
    >
      <div
        data-radium={true}
        style={
          Object {
            "position": "relative",
          }
        }
      >
        <button
          data-radium={true}
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "MozOsxFontSmoothing": "grayscale",
              "WebkitFontSmoothing": "antialiased",
              "alignItems": "center",
              "backgroundColor": "#43B02A",
              "backgroundImage": "none",
              "border": "1px solid transparent",
              "borderRadius": "4px",
              "color": "#FFFFFF",
              "cursor": "pointer",
              "display": "inline-flex",
              "fontSize": "16px",
              "fontWeight": 600,
              "height": "40px",
              "paddingLeft": 16,
              "paddingRight": 16,
              "touchAction": "manipulation",
              "userSelect": "none",
              "whiteSpace": "nowrap",
            }
          }
          tabIndex={undefined}
          type="button"
        >
           Share 
        </button>
      </div>
      <div
        data-radium={true}
        style={
          Object {
            "display": "block",
            "pointerEvents": "none",
            "position": "absolute",
            "width": "100%",
            "zIndex": -9000,
          }
        }
      >
        <div
          style={
            Object {
              "overflowX": "visible",
              "overflowY": "hidden",
              "padding": "2px 5px 5px",
              "transform": "translateX(-5px)",
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "transform": "translateY(-30px)",
                "transition": "all 200ms ease-in-out",
              }
            }
          >
            <div
              style={
                Object {
                  "opacity": 0,
                  "transition": "all 200ms ease-in-out",
                }
              }
            >
              <div
                aria-hidden={true}
                onBlur={[Function]}
                onKeyDown={[Function]}
                role="menu"
                style={
                  Object {
                    "backgroundColor": "#ffffff",
                    "borderRadius": "4px",
                    "boxShadow": "0 3px 10px 0 rgba(0, 0, 0, 0.16), 0 3px 10px 0 rgba(0, 0, 0, 0.23)",
                    "display": "inline-block",
                    "maxHeight": "500px",
                    "outline": "none",
                    "overflowY": "auto",
                    "padding": "8px 0",
                    "userSelect": "none",
                  }
                }
                tabIndex={-1}
              >
                <div
                  data-radium={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="menuitem"
                  style={
                    Object {
                      "backgroundColor": "#ffffff",
                      "padding": "16px 0",
                    }
                  }
                  tabIndex={0}
                >
                  <div
                    data-radium={true}
                    style={
                      Object {
                        "display": "table",
                      }
                    }
                  >
                    <div
                      data-radium={true}
                      style={
                        Object {
                          "display": "table-cell",
                          "paddingLeft": "16px",
                          "verticalAlign": "middle",
                        }
                      }
                    >
                      <div
                        data-radium={true}
                        style={
                          Object {
                            "display": "flex",
                          }
                        }
                      >
                        <i
                          aria-hidden={true}
                          data-radium={true}
                          onClick={undefined}
                          style={
                            Object {
                              "fontFamily": "ic-icons",
                              "fontSize": "22px",
                              "fontSmoothing": "antialiased",
                              "fontStyle": "normal",
                              "fontVariant": "normal",
                              "fontWeight": "normal",
                              "lineHeight": "1",
                              "osxFontSmoothing": "grayscale",
                              "position": "relative",
                              "speak": "none",
                              "textTransform": "none",
                            }
                          }
                        >
                          
                        </i>
                      </div>
                    </div>
                    <div
                      data-radium={true}
                      style={
                        Object {
                          "padding": "0 16px",
                          "whiteSpace": "nowrap",
                        }
                      }
                    >
                      Share via Email
                    </div>
                  </div>
                </div>
                <div
                  data-radium={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="menuitem"
                  style={
                    Object {
                      "backgroundColor": "#ffffff",
                      "padding": "16px 0",
                    }
                  }
                  tabIndex={0}
                >
                  <div
                    data-radium={true}
                    style={
                      Object {
                        "display": "table",
                      }
                    }
                  >
                    <div
                      data-radium={true}
                      style={
                        Object {
                          "display": "table-cell",
                          "paddingLeft": "16px",
                          "verticalAlign": "middle",
                        }
                      }
                    >
                      <div
                        data-radium={true}
                        style={
                          Object {
                            "display": "flex",
                          }
                        }
                      >
                        <i
                          aria-hidden={true}
                          data-radium={true}
                          onClick={undefined}
                          style={
                            Object {
                              "color": "#3c5a99",
                              "fontFamily": "ic-icons",
                              "fontSize": "22px",
                              "fontSmoothing": "antialiased",
                              "fontStyle": "normal",
                              "fontVariant": "normal",
                              "fontWeight": "normal",
                              "lineHeight": "1",
                              "osxFontSmoothing": "grayscale",
                              "position": "relative",
                              "speak": "none",
                              "textTransform": "none",
                            }
                          }
                        >
                          
                        </i>
                      </div>
                    </div>
                    <div
                      data-radium={true}
                      style={
                        Object {
                          "padding": "0 16px",
                          "whiteSpace": "nowrap",
                        }
                      }
                    >
                      Share via Facebook
                    </div>
                  </div>
                </div>
                <div
                  data-radium={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  role="menuitem"
                  style={
                    Object {
                      "backgroundColor": "#ffffff",
                      "padding": "16px 0",
                    }
                  }
                  tabIndex={0}
                >
                  <div
                    data-radium={true}
                    style={
                      Object {
                        "display": "table",
                      }
                    }
                  >
                    <div
                      data-radium={true}
                      style={
                        Object {
                          "display": "table-cell",
                          "paddingLeft": "16px",
                          "verticalAlign": "middle",
                        }
                      }
                    >
                      <div
                        data-radium={true}
                        style={
                          Object {
                            "display": "flex",
                          }
                        }
                      >
                        <i
                          aria-hidden={true}
                          data-radium={true}
                          onClick={undefined}
                          style={
                            Object {
                              "color": "#1da1f2",
                              "fontFamily": "ic-icons",
                              "fontSize": "22px",
                              "fontSmoothing": "antialiased",
                              "fontStyle": "normal",
                              "fontVariant": "normal",
                              "fontWeight": "normal",
                              "lineHeight": "1",
                              "osxFontSmoothing": "grayscale",
                              "position": "relative",
                              "speak": "none",
                              "textTransform": "none",
                            }
                          }
                        >
                          
                        </i>
                      </div>
                    </div>
                    <div
                      data-radium={true}
                      style={
                        Object {
                          "padding": "0 16px",
                          "whiteSpace": "nowrap",
                        }
                      }
                    >
                      Share via Twitter
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;
